### PR TITLE
Allow to set tracking branch in rename (closes #1484)

### LINKED ISSUE #1484: Is it ok that the rename action keeps the remote branch name?
## Steps to reproduce
1. Slide In Branch: `foo`
2. Push `foo`
3. Rename `foo` to `bar`
4. Push `bar` 

Actual:
Push to `origin/foo`.

Expected:
TBD, I was thinking that it should be `origin/bar`. 

## (optional) Screenshots
image

## Possible solutions 
1. current state is ok
2. Unset remote `origin/foo` (how? is it git config?) - then the branch renamed to the new name should be 
  - untracked - if `origin/bar` does not exist
  - any to remote - if `origin/bar` does exist 
3. A checkbox in the rename dialog so one can decide about the rename of the remote (?)

### PR DESCRIPTION
**Pull request by mkondratek**

### CODE REVIEW

File: frontend/actions/src/main/java/com/virtuslab/gitmachete/frontend/actions/base/BaseRenameAction.java [hunk 1]
```diff
@@ -82,7 +82,7 @@ private void rename(GitRepository gitRepository, BaseEnhancedGraphTable graphTab
         branch.getName(),
         /* showCheckOutOption */ false,
         /* showResetOption */ false,
-        /* showSetTrackingOption */ false,
+        /* showSetTrackingOption */ true,
```
Comment (PawelLipski) [inline frontend/actions/src/main/java/com/virtuslab/gitmachete/frontend/actions/base/BaseRenameAction.java:85]: Wow okay so the functionality is already implemented in git4idea?

File: frontend/actions/src/main/java/com/virtuslab/gitmachete/frontend/actions/backgroundables/RenameBackgroundable.java [hunk 2]
```diff
@@ -78,6 +84,57 @@ public void run(ProgressIndicator indicator) {
...
+    }
+  }
+
+  @UIThreadUnsafe
+  private void handleSetTrackingBranch(String remoteName) {
+    val newBranchName = gitNewBranchOptions.getName();
+    val renamedRemoteName = "${remoteName}/${newBranchName}";
+
+    // setUpstream: git branch --set-upstream-to <upstreamBranchName> <branchName>
+    val setUpstream = Git.getInstance().setUpstream(gitRepository, renamedRemoteName, newBranchName);
```
Comment (PawelLipski) [inline frontend/actions/src/main/java/com/virtuslab/gitmachete/frontend/actions/backgroundables/RenameBackgroundable.java:101]: Blocking call right?

File: frontend/actions/src/main/java/com/virtuslab/gitmachete/frontend/actions/backgroundables/RenameBackgroundable.java [hunk 3]
```diff
@@ -78,6 +84,57 @@ public void run(ProgressIndicator indicator) {
...
+    var error = setUpstream.getErrorOutputAsJoinedString();
+
+    if (!setUpstream.success()) {
+      if (setUpstream.getErrorOutput().get(0)
```
Comment (PawelLipski) [inline frontend/actions/src/main/java/com/virtuslab/gitmachete/frontend/actions/backgroundables/RenameBackgroundable.java:105]: Whoops I'd be careful about the case with no error output at all :frown:

File: frontend/actions/src/main/java/com/virtuslab/gitmachete/frontend/actions/backgroundables/RenameBackgroundable.java (hunk 2 above)
Comment (PawelLipski) [reply frontend/actions/src/main/java/com/virtuslab/gitmachete/frontend/actions/backgroundables/RenameBackgroundable.java:101]: Yeah, `git4idea.commands.GitImpl#setUpstream` and `git4idea.commands.GitImplBase#run(git4idea.commands.GitLineHandler, git4idea.commands.GitImplBase.OutputCollector)` look blocking

File: frontend/actions/src/main/java/com/virtuslab/gitmachete/frontend/actions/backgroundables/RenameBackgroundable.java [hunk 4]
```diff
@@ -78,6 +84,57 @@ public void run(ProgressIndicator indicator) {
...
+        }
+        error = unsetUpstream.getErrorOutputAsJoinedString();
+      }
+
+      VcsNotifier.getInstance(project).notifyError(null,
+          getString("action.GitMachete.RenameBackgroundable.rename-failed"),
+          error);
+    } else {
+      // This happens when one renames a branch with a remote branch set up,
+      // and the remote with the new name exist.
```
Comment (PawelLipski) [inline frontend/actions/src/main/java/com/virtuslab/gitmachete/frontend/actions/backgroundables/RenameBackgroundable.java:128]: Nit: do you mean:
```suggestion
      // and the remote branch with the new name exist.
```

File: frontend/actions/src/main/java/com/virtuslab/gitmachete/frontend/actions/backgroundables/RenameBackgroundable.java (hunk 3 above)
Comment (PawelLipski) [reply frontend/actions/src/main/java/com/virtuslab/gitmachete/frontend/actions/backgroundables/RenameBackgroundable.java:105]: In case `git branch` gets killed or whatnot and doesn't produce any stderr in time

File: frontend/actions/src/main/java/com/virtuslab/gitmachete/frontend/actions/backgroundables/RenameBackgroundable.java [hunk 5]
```diff
@@ -78,6 +84,57 @@ public void run(ProgressIndicator indicator) {
...
+  @UIThreadUnsafe
+  private void handleSetTrackingBranch(String remoteName) {
+    val newBranchName = gitNewBranchOptions.getName();
+    val renamedRemoteName = "${remoteName}/${newBranchName}";
+
+    // setUpstream: git branch --set-upstream-to <upstreamBranchName> <branchName>
+    val setUpstream = Git.getInstance().setUpstream(gitRepository, renamedRemoteName, newBranchName);
+    var error = setUpstream.getErrorOutputAsJoinedString();
+
+    if (!setUpstream.success()) {
```
Comment (PawelLipski) [inline frontend/actions/src/main/java/com/virtuslab/gitmachete/frontend/actions/backgroundables/RenameBackgroundable.java:104]: I'd invert this `if` to sth like

```suggestion
    if (setUpstream.success()) return;
```

File: frontend/actions/src/main/java/com/virtuslab/gitmachete/frontend/actions/backgroundables/RenameBackgroundable.java [hunk 6]
```diff
@@ -78,6 +84,57 @@ public void run(ProgressIndicator indicator) {
...
+      if (setUpstream.getErrorOutput().get(0)
+          .equals("fatal: the requested upstream branch '${renamedRemoteName}' does not exist")) {
+        val unsetUpstream = unsetUpstream(newBranchName);
+        if (unsetUpstream.success()) {
+          // This happens when one renames a branch with a remote branch set up,
+          // and a local branch with the new name is created for the second and next time,
+          // and the remote with the new name does not exist.
```
Comment (PawelLipski) [inline frontend/actions/src/main/java/com/virtuslab/gitmachete/frontend/actions/backgroundables/RenameBackgroundable.java:111]: Huh could you provide a sequence of `git` commands in this comment  that would lead to such an effect? that's pretty hard to follow 🤔

File: frontend/actions/src/main/java/com/virtuslab/gitmachete/frontend/actions/backgroundables/RenameBackgroundable.java [hunk 7]
```diff
@@ -78,6 +84,57 @@ public void run(ProgressIndicator indicator) {
     // (Stepping deeper is not an option since we would lose some important logic or become very dependent on the internals of git4idea).
     // Hence, we wait for the creation of the branch (with exponential backoff).
     waitForCreationOfLocalBranch(gitRepository, newBranchName);
+
+    if (gitNewBranchOptions.shouldSetTracking()) {
```
Comment (PawelLipski) [inline frontend/actions/src/main/java/com/virtuslab/gitmachete/frontend/actions/backgroundables/RenameBackgroundable.java:88]: Hmmm

image

1. `~/git-machete/graphics/setup-sandbox`
2. open IntelliJ on `~/machete-sandbox`
3. Rename `allow-ownership-link`
4. Whoops it's untracked now

Is it expected? 🤔

Comment (PawelLipski) [reply frontend/actions/src/main/java/com/virtuslab/gitmachete/frontend/actions/backgroundables/RenameBackgroundable.java:88]: Also:
image
No trace of `git branch --set-upstream-to` in history 🤔

File: frontend/actions/src/main/java/com/virtuslab/gitmachete/frontend/actions/base/BaseRenameAction.java [hunk 8]
```diff
@@ -76,13 +77,14 @@ public void actionPerformed(AnActionEvent anActionEvent) {
...
     val project = gitRepository.getProject();
+    val remotes = new ArrayList<>(gitRepository.getRemotes());
 
     val gitNewBranchDialog = new GitNewBranchDialog(project, Collections.singletonList(gitRepository),
         getString("action.GitMachete.BaseRenameAction.description").fmt(branch.getName()),
         branch.getName(),
         /* showCheckOutOption */ false,
         /* showResetOption */ false,
-        /* showSetTrackingOption */ false,
+        /* showSetTrackingOption */ branch.getRemoteTrackingBranch() != null,
```
Comment (PawelLipski) [inline frontend/actions/src/main/java/com/virtuslab/gitmachete/frontend/actions/base/BaseRenameAction.java:87]: Huh would it be easily possible to rename this option from _Set tracking branch_ (which is unclear) to _Rename tracking branch_?

File: frontend/actions/src/main/java/com/virtuslab/gitmachete/frontend/actions/backgroundables/RenameBackgroundable.java (hunk 4 above)
Comment (PawelLipski) [reply frontend/actions/src/main/java/com/virtuslab/gitmachete/frontend/actions/backgroundables/RenameBackgroundable.java:128]: (as _remote_ on its own typically refers to a _remote repository_, not _remote branch_)

File: frontend/actions/src/main/java/com/virtuslab/gitmachete/frontend/actions/backgroundables/RenameBackgroundable.java (hunk 2 above)
Comment (mkondratek) [reply frontend/actions/src/main/java/com/virtuslab/gitmachete/frontend/actions/backgroundables/RenameBackgroundable.java:101]: yes

File: frontend/actions/src/main/java/com/virtuslab/gitmachete/frontend/actions/base/BaseRenameAction.java (hunk 8 above)
Comment (mkondratek) [reply frontend/actions/src/main/java/com/virtuslab/gitmachete/frontend/actions/base/BaseRenameAction.java:87]: I am  gonna to copy and revamp GitNewBranchDialog to achieve that :trollface:

File: frontend/actions/src/main/java/com/virtuslab/gitmachete/frontend/actions/backgroundables/RenameBackgroundable.java (hunk 7 above)
Comment (mkondratek) [reply frontend/actions/src/main/java/com/virtuslab/gitmachete/frontend/actions/backgroundables/RenameBackgroundable.java:88]: you have not specified whether the set tracking branch checkbox is selected in your scenario.

if it is selected it is expected IMHO - we "rename" the remote branch to smth that does not exist in the remote repo yet 

in the other case the status to remote shall remain ofc ( and that's how it works to me )

File: frontend/actions/src/main/kotlin/com/virtuslab/gitmachete/frontend/actions/dialogs/MyGitNewBranchDialog.kt [hunk 9]
```diff
@@ -0,0 +1,211 @@
...
+import com.intellij.util.textCompletion.TextCompletionProviderBase
+import com.intellij.util.textCompletion.TextFieldWithCompletion
+import git4idea.branch.GitBranchOperationType
+import git4idea.branch.GitNewBranchOptions
+import git4idea.i18n.GitBundle
+import git4idea.repo.GitRepository
+import git4idea.validators.*
+import javax.swing.JCheckBox
+
+class MyGitNewBranchDialog @JvmOverloads constructor(
```
Comment (PawelLipski) [inline frontend/actions/src/main/kotlin/com/virtuslab/gitmachete/frontend/actions/dialogs/MyGitNewBranchDialog.kt:25]: Wowow pls mention the original class in the comment

File: frontend/resourcebundles/src/main/java/com/virtuslab/gitmachete/frontend/resourcebundles/GitMacheteBundle.java [hunk 10]
```diff
@@ -37,11 +39,67 @@ private GitMacheteBundle() {}
...
   @I18nMakeFormat
   public static @Untainted String getNonHtmlString(@PropertyKey(resourceBundle = BUNDLE) String key) {
     return instance.getString(key);
   }
+
+  /**
+   * This method has been inspired by {@link com.intellij.BundleBase#replaceMnemonicAmpersand}.
+   * It is required to correctly process the mnemonics.
+   */
+  private static String replaceMnemonicAmpersand(String value) {
```
Comment (mkondratek) [inline frontend/resourcebundles/src/main/java/com/virtuslab/gitmachete/frontend/resourcebundles/GitMacheteBundle.java:55]: We do not process ampersand mnemonics as we should (at least not in all cases) 😅 

I realised it while I created `MyGitNewBranchDialog` and saw `&` characters in the message strings. Then I compared our bundle class with git4Idea bundle class and found `BundleBase`. 

Options here: 
1. To keep this PR smaller - I'd rather keep the current approach for now and 
  a) just leave it as good enough
  b) try to extend `BundleBase` (or `DynamicBundle`) in `GitMacheteBundle` later
    - probably that's smth we not only can but actually should do 
    - the first disadvantage of this solution is that the resourcebundle module will need a dependency on intellij IIUC
2. To keep PR smaller - I could remove the ampersands from the string bundles  for now. And maybe add it later if we implement `BundleBase` or smth.

@PawelLipski , wdyt?

File: frontend/actions/src/main/java/com/virtuslab/gitmachete/frontend/actions/backgroundables/RenameBackgroundable.java (hunk 7 above)
Comment (PawelLipski) [reply frontend/actions/src/main/java/com/virtuslab/gitmachete/frontend/actions/backgroundables/RenameBackgroundable.java:88]: Huh okay... makes sense. Still, pretty surprising at first glance...

File: frontend/resourcebundles/src/main/java/com/virtuslab/gitmachete/frontend/resourcebundles/GitMacheteBundle.java (hunk 10 above)
Comment (PawelLipski) [reply frontend/resourcebundles/src/main/java/com/virtuslab/gitmachete/frontend/resourcebundles/GitMacheteBundle.java:55]: Huh actually... do you ever use the ampersand mnemonics yourself? 🤔 I don't remember ever using them, if I use some action frequently then I add a regular shortcut. So... (2.) sounds like a reasonable solution.

Also... what's the problem with **not** processing the mnemonics? did they stop working (assuming no `0x1B` character is added) with some version of IntelliJ? 🤔

Comment (mkondratek) [reply frontend/resourcebundles/src/main/java/com/virtuslab/gitmachete/frontend/resourcebundles/GitMacheteBundle.java:55]: > Huh actually... do you ever use the ampersand mnemonics yourself? 

yes, sometimes

File: frontend/actions/src/main/java/com/virtuslab/gitmachete/frontend/actions/backgroundables/RenameBackgroundable.java (hunk 6 above)
Comment (mkondratek) [reply frontend/actions/src/main/java/com/virtuslab/gitmachete/frontend/actions/backgroundables/RenameBackgroundable.java]: IDE git console shows `git branch -m ...` calls only.

You can simply reproduce it.
1. `./scripts/setup-multiroot-sandbox`
2. `runIde` 
3. Rename `allow-ownership-link` to `allow-ownership-link-XD` with the option to rename remote
4. Rename `allow-ownership-link-XD` to `allow-ownership-link` with the option to rename remote 
5. Rename `allow-ownership-link` to `allow-ownership-link-XD` with the option to rename remote

File: CHANGE-NOTES.md [hunk 11]
```diff
@@ -14,6 +14,7 @@
 - Added a field to provide/edit a custom annotation in the Slide In dialog.
 - Added Compare With Parent action to the context menu.
 - Fixed an issue with Slide In Unmanaged Branch Notification for a branch from a git repository that is not selected.
+- Added an option to the branch rename dialog allowing to set the remote branch name.
```
Comment (PawelLipski) [inline CHANGE-NOTES.md:17]: Maybe 
```suggestion
- Added an option to the branch rename dialog allowing to keep or discard the remote tracking branch info.
```

File: frontend/resourcebundles/src/main/resources/GitMacheteBundle.properties [hunk 12]
```diff
@@ -144,6 +146,13 @@ action.GitMachete.ResetCurrentToRemoteBackgroundable.task-title=Resetting\u2026
 
 action.GitMachete.BaseRenameAction.description=Rename branch ''{0}''
 
+string.GitMachete.NewBranchDialog.new-branch-name=New branch name:
+string.GitMachete.NewBranchDialog.checkout-checkbox=Checkout branch
+string.GitMachete.NewBranchDialog.keep-remote-checkbox=Keep remote tracking info
```
Comment (PawelLipski) [inline frontend/resourcebundles/src/main/resources/GitMacheteBundle.properties:151]: Way clearer!

**Review by PawelLipski — COMMENTED** [review]
Anything left to do (still a draft)?

**Comment by mkondratek** [comment]
> Anything left to do (still a draft)?

I am finalising some backwards compatibility tweaks - it will be ready to review soon! 

Here is `runPluginVerifier` result for the record (without the adjustments): https://app.circleci.com/pipelines/github/VirtusLab/git-machete-intellij-plugin/8915/workflows/5cebf6a3-7440-4809-91fc-c563ff14ccfd/jobs/9346

File: frontend/actions/src/main/java/com/virtuslab/gitmachete/frontend/actions/common/BranchNamesCompletion.java [hunk 13]
```diff
@@ -0,0 +1,92 @@
...
+import com.intellij.codeInsight.lookup.LookupElement;
+import com.intellij.codeInsight.lookup.LookupElementBuilder;
+import com.intellij.openapi.project.DumbAware;
+import com.intellij.util.textCompletion.DefaultTextCompletionValueDescriptor;
+import com.intellij.util.textCompletion.TextCompletionProvider;
+import com.intellij.util.textCompletion.TextCompletionValueDescriptor;
+import io.vavr.collection.List;
+import lombok.val;
+
+// TODO (#1604): remove this class and replace with com.intellij.util.textCompletion.TextCompletionProviderBase
```
Comment (PawelLipski) [inline frontend/actions/src/main/java/com/virtuslab/gitmachete/frontend/actions/common/BranchNamesCompletion.java:18]: I'm assuming that's more or less 1:1 copierd from the original source? 😅

Comment (mkondratek) [reply frontend/actions/src/main/java/com/virtuslab/gitmachete/frontend/actions/common/BranchNamesCompletion.java:18]: yes! and simplified for our single purpose

**Review by PawelLipski — APPROVED** [review]
Oh! pls also add a mention about this newly-added option to features.md + some 1/2 sentences on what happens in case the checkbox is/isn't selected

**Comment by mkondratek** [comment]
image

File: frontend/resourcebundles/src/main/resources/GitMacheteBundle.properties [hunk 14]
```diff
@@ -144,6 +146,14 @@ action.GitMachete.ResetCurrentToRemoteBackgroundable.task-title=Resetting\u2026
 
 action.GitMachete.BaseRenameAction.description=Rename branch ''{0}''
 
+string.GitMachete.MyGitNewBranchDialog.new-branch-name=New branch name:
+string.GitMachete.MyGitNewBranchDialog.checkout-checkbox=Checkout branch
+string.GitMachete.MyGitNewBranchDialog.keep-remote-checkbox=Keep remote tracking info
+string.GitMachete.MyGitNewBranchDialog.keep-remote-checkbox.tooltip.HTML=<html>If 'Keep remote tracking info' is selected the remote branch name remains<nobr/>.<br> Uncheck the box to rename the remote branch along with the local branch name.</html>
```
[the remark this answers is not in the document]
Comment (mkondratek) [reply frontend/resourcebundles/src/main/resources/GitMacheteBundle.properties:152]: Good point! Yes, that's how it works. The actual rename of the remote would rise even bigger problems (e.g. what about existing GitHub PR?)
